Source/Adobe.Target.Client/Service/RuleLoader.cs

```diff
@@ -0,0 +1,219 @@
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+#nullable enable
+namespace Adobe.Target.Client.Service
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Net;
+    using System.Net.Http;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Adobe.Target.Client.Extension;
+    using Adobe.Target.Client.Model.OnDevice;
+    using Adobe.Target.Client.Util;
+    using Adobe.Target.Delivery.Model;
+    using Microsoft.Extensions.Logging;
+    using Newtonsoft.Json;
+    using Polly;
+    using RestSharp;
+
+    internal sealed class RuleLoader
+    {
+        private const string AcceptHeader = "Accept";
+        private const string EtagHeader = "ETag";
+        private const string NoneMatchHeader = "If-None-Match";
+        private const string AcceptHeaderValue = "application/json";
+        private const string MajorVersion = "1";
+        private const string ArtifactFilename = "/rules.json";
+        private const int MaxRetries = 10;
+
+        private static readonly IReadOnlyList<HttpStatusCode> HttpStatusCodesWorthRetrying = new[]
+        {
+            HttpStatusCode.RequestTimeout,
+            HttpStatusCode.InternalServerError,
+            HttpStatusCode.BadGateway,
+            HttpStatusCode.ServiceUnavailable,
+            HttpStatusCode.GatewayTimeout,
+        };
+
+        private readonly TargetClientConfig clientConfig;
+        private readonly ILogger? logger;
+        private readonly IOnDeviceDecisioningHandler? handler;
+        private Timer? timer;
+        private volatile OnDeviceDecisioningRuleSet? latestRules;
+        private string? lastETag;
+        private bool succeeded;
+
+        internal RuleLoader(TargetClientConfig clientConfig)
+        {
+            this.clientConfig = clientConfig;
+            this.logger = this.clientConfig.Logger;
+            this.handler = this.clientConfig.OnDeviceDecisioningHandler;
+            this.SetLocalRules();
+            this.ScheduleTimer(0);
+        }
+
+        private void SetLocalRules()
+        {
+            var artifactPayload = this.clientConfig.OnDeviceArtifactPayload;
+            if (string.IsNullOrEmpty(artifactPayload))
+            {
+                return;
+            }
+
+            if (!this.DeserializeRules(artifactPayload, out var deserialized))
+            {
+                return;
+            }
+
+            Interlocked.Exchange(ref this.latestRules, deserialized);
+
+            if (!this.succeeded)
+            {
```

Comment:
What would happen to `succeeded` if we load rules multiple times. As far as I can tell `succeeded` is set to `true` once and ready callback is invoked ONLY once as well.

Reply:
Right, it's called only once by design (same as in Java SDK), and we state this in the doc:
```
        /// <summary>
        /// This is called once when on-device execution is ready
        /// </summary>
        void OnDeviceDecisioningReady();
```